Draft specification for labels

DougManuel opened the pull request:
The PR has specifications for refactoring labels to address the approach of moving labels out of rec_with_table(). Further, we have little use of labels and metadata, but this is a priority feature.

There is a clearer and more consistent approach to variable labels. I propose we should extend the 'labelled' library, and there are a few ways to do so.

DougManuel committed:
Draft specification for labels

DougManuel committed:
Add label helper for labels

yulric commented on docs/labels.qmd:
```diff
@@ -0,0 +1,166 @@
+
+## Label specifications
+
+- Follow the labelled library approach. All labelled functions should work with recodeflow.
```
I would add a sub-section here about which challenges mentioned above recodeflow will tackle and which ones are out of scope.

yulric reviewed (changes requested):
The document is pretty good but missing more text about what's in scope for the library.

DougManuel committed:
Delete pbc_variable_details.xlsx added by mistake. Update renv

DougManuel committed:
Update introduction with an introduction of worksheet. PBC example.

DougManuel commented on docs/labels.qmd:
```diff
@@ -0,0 +1,166 @@
+
+## Scope and specifications for labels and metadata
+
+Adding support for labels in recodeflow will enhance data clarity and usability by providing descriptive metadata for variables, ensuring consistent interpretation across analyses. This feature will require implementing a standardized system for label management, including functions for adding, modifying, and retrieving labels within the recodeflow framework. There are two primary uses of variabl ...
```
I modified by introducing the "framework" of worksheets and S3 objects. I am concerned this may be getting technical in the intro, but the clarity is helpful.

DougManuel commented on docs/labels.qmd:
```diff
@@ -0,0 +1,166 @@
+
+Other uses include:
+
+1. Facilitating data sharing and collaboration by providing clear documentation of variable meanings and valid values.
```
An example of missing data is added.

DougManuel committed:
Label scope - draft #2

DougManuel commented:
I wrote a new version of the scope document. It hopefully addresses the comments, but unfortunately, it requires a new review.

The main body is a higher-level scope. An extensive appendix serves as a reference and explanation. That appendix was developed discussion with a LLM. References are included.

yulric committed:
remove renv files from the index

yulric commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+
+-   **No Interference with tagged NAs**: Operations on labels will not remove or corrupt tagged NA values. For instance, if a variable contains NA(b) to indicate some missing category, setting or getting labels for that variable should not turn NA(b) into a regular NA or a numeric code. The underlying data type remains the same. Our implementation will recognize `haven`’s tagged NA class and hand ...
+
+-   **Clear specification for missing in metadata:** The specification will clarify how to denote missing values in `variables.csv` or `variable_details.csv`. For example, we may decide that within the metadata, a certain column (or perhaps a special notation in the labels) indicates missing. Some systems use an annotation like “” next to the label. We might introduce a convention such as: if a v ...
```
The sheets already have a way of representing tagged NA values by using NA::a, NA::b. Were you thinking of changing that?

DougManuel replied:
No. Well, tagged NA requires a review since the approach has gained better support in R. However, we seem still on mark. I expect greatest effort should be adding vignettes for tagged NA.

yulric commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+
+While the primary focus is on labels, the design will allow extensibility to other metadata fields such as short vs. long labels, descriptions, and notes. We will incorporate these to complement the existing variables and variable_details structure, rather than overloading the R objects with extra attributes.
+
+-   **Short and long variable labels:** Many datasets distinguish between a short label (a brief name) and a longer description. For example, a variable BMI might have a short label “Body Mass Index” and a long description “BMI, calculated as weight (kg) / \[height (m)\]\^2”. In `recodeflow`, we plan to support storing both. The variables sheet can have separate columns, e.g., label_short and lab ...
```
Examples of codebook standards and IPUMS metadata?

DougManuel replied:
I removed this statement.

yulric commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+
+-   **Documentation of expected behavior:** The package vignettes and function references will describe how labels are preserved or modified during common operations. For example, if a user uses `dplyr::mutate()` to create a new variable from labeled variables, what happens? We will specify that by default, new variables will have no label (unless the user explicitly sets one), and existing label ...
+
+-   **Preservation of metadata in transformations:** We will aim for labels to survive through data transformations as much as logically possible. For instance:
```
Can you provide code examples of how you imaging this would work for each transformation type? Would we override the  dplyr's mutate method to preserve labels? Would we have a separate function that does this?

DougManuel replied:
I really don't know! 🤷 There is increasing awareness of this challenging issue within the data science communities. I was mostly flagging the issue, and we may want to leave this for future updates.

yulric reviewed (changes requested):
This new version is way better but I would include more code examples exploring some of your statements. In addition, a lot of what you said in the appendix is actually really useful and should be brought back into the main document. Finally, there are some formatting issues that I can fix.

If you want to address my comments, I can then refactor the document to:

1. Fix the formatting issues
2. Bring the appendix text into the main document
3. Write supporting code examples

DougManuel commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+This document outlines the scope and key technical requirements for improving how **`recodeflow`** handles variable and value labels (metadata). The primary objectives are:
+
+-   **Enable reproducible, standardized metadata**: Use proven conventions from `` `haven` `` and `` `labelled` `` so that data analysts can apply labels consistently in R.
+-   **Facilitate collaboration**: Store metadata in human-readable CSV sidecars (`variables` and `variable_details`) for project teams, including those who are not R users.
```
The CSV files (variables.csv and variable_details.csv) are central to the recodeflow system, not supplementary. In most contexts, let's replace "sidecar" with "metadata files" or "metadata."

DougManuel committed:
Change 'sidecar' reference to 'metadata'

DougManuel commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+-   **Enable reproducible, standardized metadata**: Use proven conventions from `` `haven` `` and `` `labelled` `` so that data analysts can apply labels consistently in R.
+-   **Facilitate collaboration**: Store metadata in human-readable CSV sidecars (`variables` and `variable_details`) for project teams, including those who are not R users.
+-   **Support data transformations**: Preserve labels as data is pivoted, joined, or subsetted, while respecting the changing context of variables.
+-   **Plan for future needs**: Align with existing R ecosystem packages and keep open the possibility of supporting DDI, LinkML, or extended role-based metadata in subsequent releases.
```
I am not sure what I intended. I've deleted the reference.

DougManuel committed:
Delete reference to "role-based metadata".

DougManuel commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+
+## 1. Integration with the R ecosystem
+
+The label system in `recodeflow` will rely on **existing standards** from packages like `` `haven` `` and `` `labelled` ``: - Use `attr(x, "label")` to store the variable label (e.g., `"Age of respondent"`). - Use `attr(x, "labels")` (with class `` `haven`_`labelled` ``) to store **value labels** (e.g., `0 = "Censored"`, `1 = "Transplant"`). - Avoid duplicating existing functions for applying lab ...
```
I've updated this section to be more focused and clear that our preferred approach would be `haven`.

Example of updated text

```
```{r}
# Sample PBC dataset
data <- data.frame(
  id = 1:5,
  status = c(0, 1, 2, 0, 2)
)

# NOT RECOMMENDED: Base R approach
# attr(data$status, "label") <- "Patient status"
# attr(data$status, "labels") <- c("Censored" = 0, "Transplant" = 1, "Dead" = 2)

# RECOMMENDED: Using `haven` package
library(haven)
data$status <- labelled(
  data$status,
  labels = c("Censored" =.0, "Transplant" = 1, "Dead" = 2),
  label = "Patient status"
)

# ALTERNATIVELY: Using `labelled` package
library(labelled)
data <- set_variable_labels(data, status = "Patient status")
data$status <- set_value_labels(data$status,
                               "Censored" = 0,
                               "Transplant" = 1,
                               "Dead" = 2)
```

DougManuel commented on docs/labels.qmd:
````diff
@@ -0,0 +1,403 @@
+
+To illustrate label handling, consider a minimal example:
+
+``` r
````
Corrected.

DougManuel commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+
+# `labelled` package approach
+library(`labelled`)
+data <- set_variable_labels(data, status = "Patient status")
```
Agree.

DougManuel committed:
Clarified how labels will build from haven and labelled. Reinforced concept of resuing.

fixed many formatting issues.

DougManuel committed:
Revised section 7 to support the new section 3.

DougManuel commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+
+-   **Memory efficiency:** Storing labels as attributes (which are essentially small vectors or strings attached to each column) is generally lightweight. A variable label is just a single string, and value labels are a vector of label mappings which, in most cases, is much smaller than the data itself (since it only has one entry per unique code). We will ensure that these attributes are only as ...
+
+-   **Avoiding data copies:** In R, adding or modifying attributes to a vector can sometimes trigger a copy of that vector in memory. To mitigate this, `recodeflow` will try to set labels at the time of variable creation or recoding (so that attributes are added before the variable is used widely, reducing duplication in memory). Bulk operations will be used where possible – for instance, applyin ...
```
Reference added. Also clarity edits.

DougManuel commented on docs/labels.qmd:
```diff
@@ -0,0 +1,403 @@
+
+-   **Avoiding data copies:** In R, adding or modifying attributes to a vector can sometimes trigger a copy of that vector in memory. To mitigate this, `recodeflow` will try to set labels at the time of variable creation or recoding (so that attributes are added before the variable is used widely, reducing duplication in memory). Bulk operations will be used where possible – for instance, applyin ...
+
+-   **Efficient lookup and manipulation:** Functions to get or set labels will be optimized to avoid unnecessary data scanning. Retrieving a label should be near O(1) (constant time) since it can directly fetch the attribute or look up in a keyed sidecar table. Setting a label will typically be O(1) for a variable label (assign an attribute) or O(m) for m labels in a value labels vector (which is ...
```
Instead, I've strengthened the statement and added references. I could provide code, but it seems better suited to the specifications.

DougManuel committed:
Small updates to appendices

yulric committed:
revised version of labelled document'

yulric commented:
@DougManuel I've added a new file `docs/new-labels.qmd` that is a revised version of the labels scope that combined everything in the main and appendix sections of your document. Can you take a look?

DougManuel commented:
It looks good and is ready to go. Merge away.
